Fix python bindings types cases

nerwalt opened the pull request:
Python type names can be `FooBar` in some places and `FOOBAR` in others. Further, the serializer and deserializer functions will be `serialize_FOOBAR` and `deserialize_FOOBAR`. This naming misalignment causes errors, and this PR attempts to fix it.

I adjusted the example to generate both Javascript and Python bindings. 

Also, the fix I implemented is quick and possibly not ideal in the context of the overall codebase. I have made the PR a draft PR as a result. Please let me know if there is a better way to achieve what I am trying to do.

teamplayer3 commented on examples/no_std/gen-bindings/src/main.rs:
```diff
@@ -1,17 +1,29 @@
+            name: "no_std_bindings_py".into(),
+            version: "0.1.0".try_into().unwrap(),
+        },
+        python::GenerationSettings::enable_all(),
```
Good idea adding bindings for js and python

teamplayer3 commented:
Could you elaborate more, in which cases this can happen, with a small example?

nerwalt commented:
Sure thing. Also, thanks for publishing this work in the first place. I have been using this crate in some form for about a year now. I started with the JS bindings but switched to Python, which wasn't as well supported as JS initially. Thus, I created an ugly, hacky fork for what I needed. I am working on a new project and noticed that Python support here has improved, so I am giving the upstream crate a second go.  If I can help improve Python support, I will try.

I created an example branch to showcase the issue I'm attempting to resolve.  See below.  
https://github.com/nerwalt/postcard-bindgen/tree/python-issues-example/examples/no_std/no_std_bindings_py/src/no_std_bindings_py

In `des.py`, there is the following
```python
T = TypeVar("T", Protocol, Packet, A1Meta)
def deserialize(obj_type: Type[T], bytes: bytes) -> T:
    d = Deserializer(bytes)

    if obj_type is PROTOCOL:
        return cast(T, deserialize_PROTOCOL(d))
    elif obj_type is PACKET:
        return cast(T, deserialize_PACKET(d))
    elif obj_type is A_1_META:
        return cast(T, deserialize_A_1_META(d))
    else:
        raise TypeError("{} not deserializable".format(obj_type))
```

However, the types in `types.py` are the following.  Note the name difference (e.g. `PACKET` in `deserialize()` and `Packet` in the `types` module).  
```python

class Packet:
    pass

@dataclass
class Protocol:
    packet: Packet

@dataclass
class A1Meta:
    name: str
    version: u16
    payload: list[u8]

class Packet_A1(Packet, tuple[A1Meta]):

    def __new__(cls, _0: A1Meta):
        return super(Packet_A1, cls).__new__(cls, (_0,))

    def __init__(self, _0: A1Meta):
        pass

    def __str__(self) -> str:
        return "{}{}".format("Packet_A1", super().__str__())

    def __format__(self, format_spec: str) -> str:
        return super().__format__(format_spec)

    def __repr__(self) -> str:
        return super().__repr__()
```

The name issue seems to arise from the following 
```rust
// let obj_name = binding_type.inner_name().to_obj_identifier();
    let obj_name = String::from(binding_type.inner_name());
```
`to_obj_identifier()` converts the names to uppercase and snake case, which seems wrong. 

Also, there is another issue that I fixed manually in the above example.  If you look at the previous commit of my issue example branch, you will see `types.py` as listed below.  The second issue is definition order (e.g. `Packet` is used in `Protocol`, but defined after `Protocol`). I think a [topological sort](https://en.wikipedia.org/wiki/Topological_sorting) of the structs/types before generation will address this, but I am not sure how to implement it yet.  There are probably other ways to address this too.
```python
@dataclass
class Protocol:
    packet: Packet
class Packet:
    pass

class Packet_A1(Packet, tuple[A1Meta]):

    def __new__(cls, _0: A1Meta):
        return super(Packet_A1, cls).__new__(cls, (_0,))

    def __init__(self, _0: A1Meta):
        pass

    def __str__(self) -> str:
        return "{}{}".format("Packet_A1", super().__str__())

    def __format__(self, format_spec: str) -> str:
        return super().__format__(format_spec)

    def __repr__(self) -> str:
        return super().__repr__()
@dataclass
class A1Meta:
    name: str
    version: u16
    payload: list[u8]
```

teamplayer3 commented on postcard-bindgen-core/src/code_gen/python/des.rs:
```diff
@@ -91,7 +91,8 @@ pub fn gen_des_functions(bindings: impl AsRef<[BindingType]>) -> Tokens {
 
 fn gen_des_function_for_type(binding_type: &BindingType) -> Tokens {
-    let obj_name = binding_type.inner_name().to_obj_identifier();
+    // let obj_name = binding_type.inner_name().to_obj_identifier();
+    let obj_name = String::from(binding_type.inner_name());
     let des_body = binding_type.gen_des_body();
     quote! {
         def deserialize_$(&obj_name)(d) -> $obj_name:
```
```suggestion
    let obj_name = binding_type.inner_name().to_obj_identifier();
    let return_type = binding_type.inner_name().to_owned();
    let des_body = binding_type.gen_des_body();
    quote! {
        def deserialize_$obj_name(d) -> $return_type:
```

teamplayer3 commented on postcard-bindgen-core/src/code_gen/python/des.rs:
```diff
@@ -136,7 +137,8 @@ pub fn gen_deserialize_func(tys: impl AsRef<[BindingType]>) -> Tokens {
 fn gen_des_case(define: &BindingType) -> (Tokens, Tokens) {
     let name = define.inner_name();
-    let type_name = name.to_obj_identifier();
+    // let type_name = name.to_obj_identifier();
+    let type_name = String::from(name);
     (
         quote!(obj_type is $(&type_name)),
         quote!(return cast(T, deserialize_$type_name(d))),
```
```suggestion
    let obj_name = name.to_obj_identifier();
    (
        quote!(obj_type is $(&type_name)),
        quote!(return cast(T, deserialize_$obj_name(d))),
```

teamplayer3 commented on postcard-bindgen-core/src/code_gen/python/des.rs:
```diff
@@ -6,7 +6,7 @@ use crate::{
     },
     registry::BindingType,
-    utils::StrExt,
+    // utils::StrExt,
```
```suggestion
    utils::StrExt,
```

teamplayer3 commented:
While digging into the problem, I came up with the problem, when having two structs with the same name in different mods in rust, that two exact same named bindings will be generated #52. Maybe you have a suggestion how to fix.

nerwalt commented:
I removed the String usage and implemented some of your suggestions.  I also did a few other cleanups.  This PR seems good for addressing the type naming errors.   I will look into definition order and [#52 ](https://github.com/teamplayer3/postcard-bindgen/issues/52) on a different branch.

teamplayer3 commented:
Closing in favor of #55. Added all your @nerwalt suggestions there.
